File: packages/sdk/typescript/human-protocol-sdk/src/encryption.ts
import * as openpgp from 'openpgp';
import { IKeyPair } from './interfaces';

/**
 * Type representing the data type of a message.
 * It can be either a string or a Uint8Array.
 */
type MessageDataType = string | Uint8Array;

function makeMessageDataBinary(message: MessageDataType): Uint8Array {
  if (typeof message === 'string') {
    return Buffer.from(message);
  }

  return message;
}

/**
 * Class for signing and decrypting messages.
 *
 * The algorithm includes the implementation of the [PGP encryption algorithm](https://github.com/openpgpjs/openpgpjs) multi-public key encryption on typescript, and uses the vanilla [ed25519](https://en.wikipedia.org/wiki/EdDSA#Ed25519) implementation Schnorr signature for signatures and [curve25519](https://en.wikipedia.org/wiki/Curve25519) for encryption. [Learn more](https://wiki.polkadot.network/docs/learn-cryptography).
 *
 * To get an instance of this class, initialization is recommended using the static [`build`](/ts/classes/Encryption/#build) method.
 */
export class Encryption {
  private privateKey: openpgp.PrivateKey;

  /**
   * Constructor for the Encryption class.
   *
   * @param privateKey - The private key.
   */
  constructor(privateKey: openpgp.PrivateKey) {
    this.privateKey = privateKey;
  }

  /**
   * Builds an Encryption instance by decrypting the private key from an encrypted private key and passphrase.
   *
   * @param privateKeyArmored - The encrypted private key in armored format.
   * @param passphrase - The passphrase for the private key (optional).
   * @returns The Encryption instance.
   *
   * @example
   * ```ts
   * import { Encryption } from '@human-protocol/sdk';
   *
   * const privateKey = 'Armored_priv_key';
   * const passphrase = 'example_passphrase';
   * const encryption = await Encryption.build(privateKey, passphrase);
   * ```
   */
  public static async build(
    privateKeyArmored: string,
    passphrase?: string
  ): Promise<Encryption> {
    const options = {
      armoredKey: privateKeyArmored,
    };

    if (!passphrase) {
      return new Encryption(await openpgp.readPrivateKey(options));
    }

    const privateKey = await openpgp.readPrivateKey(options);
    return new Encryption(
      await openpgp.decryptKey({
        privateKey,
        passphrase,
      })
    );
  }

  /**
   * This function signs and encrypts a message using the private key used to initialize the client and the specified public keys.
   *
   * @param message - Message to sign and encrypt.
   * @param publicKeys - Array of public keys to use for encryption.
   * @returns Message signed and encrypted.
   *
   * @example
   * ```ts
   * const publicKey1 = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   * const publicKey2 = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   *
   * const publicKeys = [publicKey1, publicKey2];
   * const resultMessage = await encryption.signAndEncrypt('message', publicKeys);
   * console.log('Encrypted message:', resultMessage);
   * ```
   */
  public async signAndEncrypt(
    message: MessageDataType,
    publicKeys: string[]
  ): Promise<string> {
    const pgpPublicKeys = await Promise.all(
      publicKeys.map((armoredKey) => openpgp.readKey({ armoredKey }))
    );

    const pgpMessage = await openpgp.createMessage({
      binary: makeMessageDataBinary(message),
    });
    const encrypted = await openpgp.encrypt({
      message: pgpMessage,
      encryptionKeys: pgpPublicKeys,
      signingKeys: this.privateKey,
      format: 'armored',
    });

    return encrypted as string;
  }

  /**
   * This function decrypts messages using the private key. In addition, the public key can be added for signature verification.
   *
   * @param message - Message to decrypt.
   * @param publicKey - Public key used to verify signature if needed (optional).
   * @returns Message decrypted.
   * @throws Error If signature could not be verified when public key is provided
   *
   * @example
   * ```ts
   * const publicKey = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   *
   * const resultMessage = await encryption.decrypt('message', publicKey);
   * console.log('Decrypted message:', resultMessage);
   * ```
   */
  public async decrypt(
    message: string,
    publicKey?: string
  ): Promise<Uint8Array> {
    const pgpMessage = await openpgp.readMessage({
      armoredMessage: message,
    });

    const decryptionOptions: openpgp.DecryptOptions = {
      message: pgpMessage,
      decryptionKeys: this.privateKey,
      format: 'binary',
    };

    const shouldVerifySignature = !!publicKey;
    if (shouldVerifySignature) {
      const pgpPublicKey = await openpgp.readKey({ armoredKey: publicKey });
      decryptionOptions.verificationKeys = pgpPublicKey;
    }

    const { data: decrypted, signatures } =
      await openpgp.decrypt(decryptionOptions);

    /**
     * There is an option to automatically verify signatures - `expectSigned`,
     * but atm it has a bug - https://github.com/openpgpjs/openpgpjs/issues/1803,
     * so we have to verify it manually till it's fixed.
     */
    try {
      if (shouldVerifySignature) {
        await signatures[0].verified;
      }
    } catch {
      throw new Error('Signature could not be verified');
    }

    return decrypted as Uint8Array;
  }

  /**
   * This function signs a message using the private key used to initialize the client.
   *
   * @param message - Message to sign.
   * @returns Message signed.
   *
   * @example
   * ```ts
   * const resultMessage = await encryption.sign('message');
   * console.log('Signed message:', resultMessage);
   * ```
   */
  public async sign(message: string): Promise<string> {
    const unsignedMessage = await openpgp.createCleartextMessage({
      text: message,
    });
    const cleartextMessage = await openpgp.sign({
      message: unsignedMessage,
      signingKeys: this.privateKey,
      format: 'armored',
    });

    return cleartextMessage;
  }
}

/**
 * Utility class for encryption-related operations.
 *
 * @example
 * ```ts
 * import { EncryptionUtils } from '@human-protocol/sdk';
 *
 * const publicKey = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
 * const isValid = await EncryptionUtils.verify('message', publicKey);
 * console.log('Signature valid:', isValid);
 * ```
 */
export class EncryptionUtils {
  /**
   * This function verifies the signature of a signed message using the public key.
   *
   * @param message - Message to verify.
   * @param publicKey - Public key to verify that the message was signed by a specific source.
   * @returns True if verified. False if not verified.
   *
   * @example
   * ```ts
   * const publicKey = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   * const result = await EncryptionUtils.verify('message', publicKey);
   * console.log('Verification result:', result);
   * ```
   */
  public static async verify(
    message: string,
    publicKey: string
  ): Promise<boolean> {
    const pgpPublicKey = await openpgp.readKey({ armoredKey: publicKey });
    const signedMessage = await openpgp.readCleartextMessage({
      cleartextMessage: message,
    });

    const verificationResult = await signedMessage.verify([pgpPublicKey]);
    const { verified } = verificationResult[0];

    try {
      return await verified;
    } catch {
      return false;
    }
  }

  /**
   * This function gets signed data from a signed message.
   *
   * @param message - Message.
   * @returns Signed data.
   * @throws Error If data could not be extracted from the message
   *
   * @example
   * ```ts
   * const signedData = await EncryptionUtils.getSignedData('message');
   * console.log('Signed data:', signedData);
   * ```
   */
  public static async getSignedData(message: string): Promise<string> {
    const signedMessage = await openpgp.readCleartextMessage({
      cleartextMessage: message,
    });

    try {
      return signedMessage.getText();
    } catch (e) {
      throw new Error('Could not get data: ' + e.message);
    }
  }

  /**
   * This function generates a key pair for encryption and decryption.
   *
   * @param name - Name for the key pair.
   * @param email - Email for the key pair.
   * @param passphrase - Passphrase to encrypt the private key (optional, defaults to empty string).
   * @returns Key pair generated.
   *
   * @example
   * ```ts
   * const name = 'YOUR_NAME';
   * const email = 'YOUR_EMAIL';
   * const passphrase = 'YOUR_PASSPHRASE';
   * const keyPair = await EncryptionUtils.generateKeyPair(name, email, passphrase);
   * console.log('Public key:', keyPair.publicKey);
   * ```
   */
  public static async generateKeyPair(
    name: string,
    email: string,
    passphrase = ''
  ): Promise<IKeyPair> {
    const { privateKey, publicKey, revocationCertificate } =
      await openpgp.generateKey({
        type: 'ecc',
        curve: 'ed25519Legacy',
        userIDs: [{ name: name, email: email }],
        passphrase: passphrase,
        format: 'armored',
      });

    return {
      passphrase: passphrase,
      privateKey,
      publicKey,
      revocationCertificate,
    };
  }

  /**
   * This function encrypts a message using the specified public keys.
   *
   * @param message - Message to encrypt.
   * @param publicKeys - Array of public keys to use for encryption.
   * @returns Message encrypted.
   *
   * @example
   * ```ts
   * const publicKey1 = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   * const publicKey2 = '-----BEGIN PGP PUBLIC KEY BLOCK-----...';
   * const publicKeys = [publicKey1, publicKey2];
   * const encryptedMessage = await EncryptionUtils.encrypt('message', publicKeys);
   * console.log('Encrypted message:', encryptedMessage);
   * ```
   */
  public static async encrypt(
    message: MessageDataType,
    publicKeys: string[]
  ): Promise<string> {
    const pgpPublicKeys = await Promise.all(
      publicKeys.map((armoredKey) => openpgp.readKey({ armoredKey }))
    );

    const pgpMessage = await openpgp.createMessage({
      binary: makeMessageDataBinary(message),
    });
    const encrypted = await openpgp.encrypt({
      message: pgpMessage,
      encryptionKeys: pgpPublicKeys,
      format: 'armored',
    });

    return encrypted as string;
  }

  /**
   * Verifies if a message appears to be encrypted with OpenPGP.
   *
   * @param message - Message to verify.
   * @returns `true` if the message appears to be encrypted, `false` if not.
   *
   * @example
   * ```ts
   * const message = '-----BEGIN PGP MESSAGE-----...';
   * const isEncrypted = EncryptionUtils.isEncrypted(message);
   *
   * if (isEncrypted) {
   *   console.log('The message is encrypted with OpenPGP.');
   * } else {
   *   console.log('The message is not encrypted with OpenPGP.');
   * }
   * ```
   */
  public static isEncrypted(message: string): boolean {
    const startMarker = '-----BEGIN PGP MESSAGE-----';
    const endMarker = '-----END PGP MESSAGE-----';

    const hasStartMarker = message.includes(startMarker);
    const hasEndMarker = message.includes(endMarker);

    return hasStartMarker && hasEndMarker;
  }
}

File: packages/sdk/python/human-protocol-sdk/human_protocol_sdk/encryption/encryption.py
"""Encrypt, decrypt, sign, and verify messages using PGP."""

from typing import Optional, List, Union
from pgpy import PGPKey, PGPMessage
from pgpy.constants import SymmetricKeyAlgorithm
from pgpy.errors import PGPError


class Encryption:
    """Encryption and decryption helper using PGP (Pretty Good Privacy).

    This class provides methods to sign, encrypt, decrypt, and verify messages
    using PGP encryption with private/public key pairs.

    Attributes:
        private_key (PGPKey): The unlocked PGP private key.
        passphrase (Optional[str]): Passphrase used to unlock the private key.
    """

    def __init__(self, private_key_armored: str, passphrase: Optional[str] = None):
        """Initialize an Encryption instance with a private key.

        Args:
            private_key_armored (str): Armored representation of the PGP private key.
            passphrase (Optional[str]): Passphrase to unlock the private key if it's locked.

        Raises:
            ValueError: If the private key is invalid, cannot be unlocked with the passphrase,
                or is locked and no passphrase is provided.

        Example:
            ```python
            from human_protocol_sdk.encryption import Encryption

            encryption = Encryption(
                "-----BEGIN PGP PRIVATE KEY BLOCK-----...",
                "your-passphrase"
            )
            ```
        """
        try:
            self.private_key, _ = PGPKey.from_blob(private_key_armored)
        except PGPError as e:
            raise ValueError("Invalid private key: {}".format(str(e)))

        if not self.private_key.is_unlocked:
            if passphrase:
                try:
                    with self.private_key.unlock(passphrase):
                        self.passphrase = passphrase

                except PGPError as e:
                    raise ValueError("Failed to unlock private key: {}".format(str(e)))
            else:
                raise ValueError("Private key locked. Passphrase needed")

    def sign_and_encrypt(
        self, message: Union[str, bytes], public_keys: List[str]
    ) -> str:
        """Sign and encrypt a message with recipient public keys.

        Signs the message with the private key and encrypts it for all specified recipients.

        Args:
            message (Union[str, bytes]): Message content to sign and encrypt.
            public_keys (List[str]): List of armored PGP public keys of the recipients.

        Returns:
            Armored, signed, and encrypted PGP message.

        Raises:
            ValueError: If the private key cannot be unlocked or encryption fails.

        Example:
            ```python
            from human_protocol_sdk.encryption import Encryption

            encryption = Encryption("-----BEGIN PGP PRIVATE KEY BLOCK-----...", "passphrase")
            encrypted_message = encryption.sign_and_encrypt(
                "your message",
                ["-----BEGIN PGP PUBLIC KEY BLOCK-----..."],
            )
            ```
        """

        pgp_message = PGPMessage.new(message)

        if not self.private_key.is_unlocked:
            try:
                with self.private_key.unlock(self.passphrase):
                    pgp_message |= self.private_key.sign(pgp_message)
            except PGPError as e:
                raise ValueError("Failed to unlock private key: {}".format(str(e)))
        else:
            pgp_message |= self.private_key.sign(pgp_message)

        cipher = SymmetricKeyAlgorithm.AES256
        sessionkey = cipher.gen_key()

        for public_key in public_keys:
            recipient_key, _ = PGPKey.from_blob(public_key)
            pgp_message = recipient_key.encrypt(pgp_message, sessionkey)

        del sessionkey
        return pgp_message.__str__()

    def decrypt(self, message: str, public_key: Optional[str] = None) -> bytes:
        """Decrypt a message using the private key.

        Decrypts an encrypted message and optionally verifies the signature using
        the sender's public key.

        Args:
            message (str): Armored PGP message to decrypt.
            public_key (Optional[str]): Optional armored public key to verify the message signature.

        Returns:
            Decrypted message as bytes.

        Raises:
            ValueError: If the private key cannot be unlocked, decryption fails,
                or signature verification fails when a public key is provided.

        Example:
            ```python
            from human_protocol_sdk.encryption import Encryption

            encryption = Encryption("-----BEGIN PGP PRIVATE KEY BLOCK-----...", "passphrase")
            decrypted_message = encryption.decrypt(encrypted_message)

            # Or with signature verification:
            decrypted_message = encryption.decrypt(
                encrypted_message,
                public_key="-----BEGIN PGP PUBLIC KEY BLOCK-----..."
            )
            ```
        """
        pgp_message = PGPMessage.from_blob(message)
        decrypted_message = ""
        try:
            if not self.private_key.is_unlocked:
                try:
                    with self.private_key.unlock(self.passphrase):
                        decrypted_message = self.private_key.decrypt(pgp_message)
                except PGPError as e:
                    raise ValueError("Failed to unlock private key: {}".format(str(e)))
            else:
                decrypted_message = self.private_key.decrypt(pgp_message)
            if public_key:
                public_key, _ = PGPKey.from_blob(public_key)
                public_key.verify(decrypted_message)

            if isinstance(decrypted_message.message, str):
                return bytes(decrypted_message.message, encoding="utf-8")
            else:
                return bytes(decrypted_message.message)
        except PGPError as e:
            if (
                decrypted_message
                and len(decrypted_message.signatures) > 0
                and str(e) == "No signatures to verify"
            ):
                raise ValueError(
                    "Failed to decrypt message: Could not find signature with this public key"
                )
            raise ValueError("Failed to decrypt message: {}".format(str(e)))

    def sign(self, message: Union[str, bytes]) -> str:
        """Sign a message with the private key.

        Creates a cleartext signed message that can be verified by anyone with
        the corresponding public key.

        Args:
            message (Union[str, bytes]): Message content to sign.

        Returns:
            Armored signed PGP message in cleartext format.

        Raises:
            ValueError: If the private key cannot be unlocked or signing fails.

        Example:
            ```python
            from human_protocol_sdk.encryption import Encryption

            encryption = Encryption("-----BEGIN PGP PRIVATE KEY BLOCK-----...", "passphrase")
            signed_message = encryption.sign("MESSAGE")
            ```
        """
        message = PGPMessage.new(message, cleartext=True)
        if not self.private_key.is_unlocked:
            try:
                with self.private_key.unlock(self.passphrase):
                    message |= self.private_key.sign(message)
            except PGPError as e:
                raise ValueError("Failed to unlock private key: {}".format(str(e)))
        else:
            message |= self.private_key.sign(message)
        return message.__str__()
